### ISSUE TITLE
Metric Cache seems to get inconsistent when state variable is a list

### ISSUE THREAD

**Issue by daMichaelB** [PyTorchLightning/metrics#766]
## 🐛 Bug

I debugged into this for a longer time now. I am not 100% sure if it is a bug, but i have no idea what i am doing wrong.

I have attached my custom metric to the `Trainer` and call `trainer.validate(model, datamodule)` to compute the metric on the whole Validation set. In each step i print my metric in the `.compute()` function. 

What confused me a lot is, that the `val_mymetric` value is always the same as the metric value of the last batch. Heres my output (only two batches):

```python
mymetric after 1. batch: 0.109375
mymetric after 2. batch: 0.20454546809196472
# end output of .validate():
{'val_cross_entropy': 0.2601596713066101,
 'val_mymetric': 0.20454546809196472}
```

I also found that in the last computation the metric only holds the values from the last batch in `preds`

### My Metric (only important functions)

```python
class MyMetric(Metric, metaclass=abc.ABCMeta):
    def __init__(self):
        super(MyMetric, self).__init__()
        self.softmax = nn.Softmax(dim=1)
        self.add_state("preds", list())
        self.add_state("labels", list())

    def reset(self):
        self.preds.clear()
        self.labels.clear()

    def update(self, preds: torch.Tensor, target: torch.Tensor):
        self.preds.append(self.softmax(preds))
        self.labels.append(target)
```

IMHO the problem is the following:

```python
# metric.py

# save context before switch
cache = {attr: getattr(self, attr) for attr in self._defaults}
```

The cache is taking the exact same object that get's cleared by the `.reset()` function. I think the cache should be built with a copy of each object? The `id()` of the cache object and metric object was the same.

### To Reproduce

as the metric is confidential i am not able to add a bigger code snippet here.

### Expected behavior

At the end of the epoch, the Cache-Content should hold all elements that come in by `update()` and not be cleared by the `.reset()`

### Environment

PyTorch Version (e.g., 1.0): 1.10.0
Flash Version: 0.6.0
OS (e.g., Linux): Ubuntu on Docker
How you installed PyTorch (conda, pip, source): pip with virtualenv
Build command you used (if compiling from source):
Python version: 3.8.12
CUDA/cuDNN version: 11.4
torchmetrics: 0.6.2

**Comment by daMichaelB** [comment]
Creating a new list-object instead of clearing the existing one seems to fix it. The list in cache is then a different object then in the metric. Maybe worth mentioning in docs?

```python
def reset(self):
    self.preds = []
    self.labels = []
```

**Comment by Borda** [comment]
@daMichaelB what TorchMetrics version are you using? :rabbit:

**Comment by daMichaelB** [comment]
> @daMichaelB what TorchMetrics version are you using? 🐰

Hey @Borda: it is version 0.6.2

**Comment by Borda** [comment]
> Creating a new list-object instead of clearing the existing one seems to fix it. The list in cache is then a different object then in the metric. Maybe worth mentioning in docs?

mind send a PR?
cc: @SkafteNicki

**Comment by daMichaelB** [comment]
> > Creating a new list-object instead of clearing the existing one seems to fix it. The list in cache is then a different object then in the metric. Maybe worth mentioning in docs?
> 
> mind send a PR? cc: @SkafteNicki

Sure! I will add a PR to mention it in the docs

**Comment by SkafteNicki** [comment]
Hi @daMichaelB,
Maybe I am misunderstanding something here, but normally we do not recommend that users implement their own `reset` method. If all states are initialized with `self.add_state` the default `reset` method will reset them for you:
https://github.com/PyTorchLightning/metrics/blob/43a226101e05ec8e5a5fe4ad64b90c44af1a3b34/torchmetrics/metric.py#L396-L409
In the case of list it will automatically create a new one.

**Comment by daMichaelB** [comment]
Hey @SkafteNicki . Oh damn, that explains a lot. Thanks for clarification. then i misunderstood this 😄 
I guess we can then close it ?! Thanks!

**Comment by SkafteNicki** [comment]
Hi @daMichaelB ,
No I think this is a clear indication that we do not communicate well enough that `reset` method should not be implemented by the user. So let's keep it and figure out a doc fix :]

**Comment by Borda** [comment]
@daMichaelB could have an idea how to reformulate it... mind sending your proposal as a PR?

**Comment by SkafteNicki** [comment]
@daMichaelB could you take a look at PR #785 if that improves the documentation on this issue?

**Comment by daMichaelB** [comment]
Hey @SkafteNicki . Thats perfect! thank you very much for the great work here! I think that makes it very clear
